Implement CQL2-JSON AST for datetime index selection

YuriZmytrakov opened the pull request:
This PR implements a CQL2 Abstract Syntax Tree (AST) structure for CQL2 Filters to enable better parameter access for index selection in search queries: 
- Introduces a structured AST representation of CQL2 queries (replacing dict-based processing)
- Adds datetime extraction logic to easily identify and access datetime parameters
- Provides ability for selecting appropriate indexes such as datetime, start_datetime, end_datetime
The AST-based approach makes it easier to parse CQL2 filters, particularly for identifying datetime constraints (datetime, start_datetime, end_datetime) to optimize search performance through proper index selection.

Examples of AST nodes:
AND:
```
LogicalNode(op=<LogicalOp.AND: 'and'>, children=[ComparisonNode(op=<ComparisonOp.EQ: '='>, field='collection', value='landsat-8-l1'), 
                                                 ComparisonNode(op=<ComparisonOp.GT: '>'>, field='properties.cloud_cover', value=50)])
```

OR:
```
LogicalNode(op=<LogicalOp.OR: 'or'>, children=[ComparisonNode(op=<ComparisonOp.EQ: '='>, field='properties.platform', value='landsat-8'), 
                                               ComparisonNode(op=<ComparisonOp.EQ: '='>, field='properties.platform', value='sentinel-2')])
```

NOT:
```
LogicalNode(op=<LogicalOp.NOT: 'not'>, children=[ComparisonNode(op=<ComparisonOp.EQ: '='>, field='properties.cloud_cover', value=0)])
```

EQ (=):
```
ComparisonNode(op=<ComparisonOp.EQ: '='>, field='properties.cloud_cover', value=20)
```

NEQ (<>)
```
ComparisonNode(op=<ComparisonOp.NEQ: '<>'>, field='properties.platform', value='landsat-8')
```

LT (<)
```
ComparisonNode(op=<ComparisonOp.LT: '<'>, field='properties.cloud_cover', value=30)
```

LTE (<=)
```
ComparisonNode(op=<ComparisonOp.LTE: '<='>, field='properties.cloud_cover', value=50)
```

GT (>)
```
ComparisonNode(op=<ComparisonOp.GT: '>'>, field='properties.gsd', value=10)
```

GTE (>=)
```
ComparisonNode(op=<ComparisonOp.GTE: '>='>, field='properties.cloud_cover', value=0)
```

IS_NULL
```
ComparisonNode(op=<ComparisonOp.IS_NULL: 'isNull'>, field='properties.cloud_cover', value=None)
```

LIKE
```
AdvancedComparisonNode(op=<AdvancedComparisonOp.LIKE: 'like'>, field='id', value='LC08%')
```

BETWEEN
```
AdvancedComparisonNode(op=<AdvancedComparisonOp.BETWEEN: 'between'>, field='properties.cloud_cover', value=(10, 50))
```

IN
```
AdvancedComparisonNode(op=<AdvancedComparisonOp.IN: 'in'>, field='properties.platform', value=['landsat-8', 'sentinel-2', 'modis'])
```

S_INTERSECTS
```
SpatialNode(op=<SpatialOp.S_INTERSECTS: 's_intersects'>, field='geometry', geometry={'type': 'Polygon', 'coordinates': [[[-110, 40], [-110, 41], [-109, 41], [-109, 40], [-110, 40]]]})
```

S_CONTAINS
```
SpatialNode(op=<SpatialOp.S_CONTAINS: 's_contains'>, field='geometry', geometry={'type': 'Point', 'coordinates': [-110.5, 40.5]})
```

S_WITHIN
```
SpatialNode(op=<SpatialOp.S_WITHIN: 's_within'>, field='geometry', geometry={'type': 'Polygon', 'coordinates': [[[-120, 35], [-120, 45], [-100, 45], [-100, 35], [-120, 35]]]})
```

S_DISJOINT
```
SpatialNode(op=<SpatialOp.S_DISJOINT: 's_disjoint'>, field='geometry', geometry={'type': 'Polygon', 'coordinates': [[[0, 0], [0, 10], [10, 10], [10, 0], [0, 0]]]})
```

DateTimeRange
```
LogicalNode(op=<LogicalOp.AND: 'and'>, children=[ComparisonNode(op=<ComparisonOp.GTE: '>='>, field='properties.datetime', value='2024-01-01T00:00:00Z'), 
                                                 ComparisonNode(op=<ComparisonOp.LTE: '<='>, field='properties.datetime', value='2024-12-31T23:59:59Z')])
```

DateTimeExactNode
```
ComparisonNode(op=<ComparisonOp.EQ: '='>, field='properties.datetime', value='2024-06-15T10:30:00Z')
```

**PR Checklist:**

- [x] Code is formatted and linted (run `pre-commit run --all-files`)
- [x] Tests pass (run `make test`)
- [x] Documentation has been updated to reflect changes, if applicable
- [x] Changes are added to the changelog

Gomez324 commented:
Hey, I ran one test with this query and body: 
127.0.0.1:8080/v1/search
```
{
  "filter-lang": "cql2-json",
  "filter": {
    "op": "and",
    "args": [
      {
        "op": "=",
        "args": [{ "property": "collection" }, "sentinel-2-l2a"]
      },
      {
        "op": ">=",
        "args": [{ "property": "datetime" }, "2015-01-01T00:00:00Z"]
      },
      {
        "op": "<=",
        "args": [{ "property": "datetime" }, "2024-12-31T23:59:59Z"]
      }
    ]
  }
}
```

And I got that index_param has this value. iit looks like it's not selecting indices. Please add tests for temporal indices.
```
(Pdb) index_param
items_*,-*kibana*,-collections*

YuriZmytrakov commented:
> Hey, I ran one test with this query and body: 127.0.0.1:8080/v1/search
> 
> ```
> {
>   "filter-lang": "cql2-json",
>   "filter": {
>     "op": "and",
>     "args": [
>       {
>         "op": "=",
>         "args": [{ "property": "collection" }, "sentinel-2-l2a"]
>       },
>       {
>         "op": ">=",
>         "args": [{ "property": "datetime" }, "2015-01-01T00:00:00Z"]
>       },
>       {
>         "op": "<=",
>         "args": [{ "property": "datetime" }, "2024-12-31T23:59:59Z"]
>       }
>     ]
>   }
> }
> ```
> 
> And I got that index_param has this value. it looks like it's not selecting temporal indices. Please add tests for temporal indices.
> 
> ```
> (Pdb) index_param
> 'items_*,-*kibana*,-collections*'
> ```

@Gomez324  I have fixed selection of collection IDs from the CQL2-JSON query, only then are the indexes selected correctly, it needs to have as input variables `collection_ids` + `datetime`. For the example query above, I can see that the following index is correctly selected:
```
(Pdb) index_param
'items_sentinel-2-l2a_2015-07-04-2017-04-16'
```

I’ve also added a test for the `index_select` function.

YuriZmytrakov commented:
> Hey, I ran one test with this query and body: 127.0.0.1:8080/v1/search
> 
> ```
> {
>   "filter-lang": "cql2-json",
>   "filter": {
>     "op": "and",
>     "args": [
>       {
>         "op": "=",
>         "args": [{ "property": "collection" }, "sentinel-2-l2a"]
>       },
>       {
>         "op": ">=",
>         "args": [{ "property": "datetime" }, "2015-01-01T00:00:00Z"]
>       },
>       {
>         "op": "<=",
>         "args": [{ "property": "datetime" }, "2024-12-31T23:59:59Z"]
>       }
>     ]
>   }
> }
> ```
> 
> And I got that index_param has this value. it looks like it's not selecting temporal indices. Please add tests for temporal indices.
> 
> ```
> (Pdb) index_param
> 'items_*,-*kibana*,-collections*'
> ```

I have added a fix for complex queries that include AND, OR, IN, NOT, BETWEEN etc. The example above now selects the indexes:
**index_param**:  
`
items_sentinel-2-l2a_2017-04-17-2018-01-22,items_sentinel-2-l2a_2018-07-20-2019-02-03,items_sentinel-2-l2a_2021-07-09-2022-01-06,items_sentinel-2-l2a_2018-01-23-2018-07-19,items_sentinel-2-l2a_2022-01-07-2022-06-30,items_sentinel-2-l2a_2020-02-02-2020-07-16,items_sentinel-2-l2a_2019-02-04-2019-07-23,items_sentinel-2-l2a_2021-01-23-2021-07-08,items_sentinel-2-l2a_2020-07-17-2021-01-22,items_sentinel-2-l2a_2022-07-01-2022-12-31,items_sentinel-2-l2a_2015-07-04-2017-04-16,items_sentinel-2-l2a_2019-07-24-2020-02-01
`

YuriZmytrakov commented:
> I created a somewhat more complex filter:
> 
> ```
> {
>     "filter-lang": "cql2-json",
>     "filter": {
>         "op": "or",
>         "args": [
>             {
>                 "op": "and",
>                 "args": [
>                     { "op": "=", "args": [{"property": "collection"}, "sentinel-2-l2a"] },
>                     { "op": ">=", "args": [{"property": "datetime"}, "2010-01-01T00:00:00Z"] },
>                     { "op": "<=", "args": [{"property": "datetime"}, "2025-12-31T23:59:59Z"] }
>                 ]
>             },
>             {
>                 "op": "and",
>                 "args": [
>                     { "op": "=", "args": [{"property": "collection"}, "sentinel-2-l1c"] },
>                     { "op": ">=", "args": [{"property": "datetime"}, "2012-01-01T00:00:00Z"] },
>                     { "op": "<=", "args": [{"property": "datetime"}, "2025-12-31T23:59:59Z"] }
>                 ]
>             }
>         ]
>     }
> }
> ```
> 
> so sentinel-2-l2a from 2010-2025 and sentinel-2-l1c from 2012-2025.
> 
> But when I debug, I see:
> 
> (Pdb) datetime_search {'gte': '2012-01-01T00:00:00.0Z', 'lte': '2025-12-31T23:59:59.0Z'}
> 
> So the filter is being parsed incorrectly. it's merging the datetime ranges instead of keeping them separate per collection.

This was resolved. Going forward, the parser can distinguish that each collection can have its own datetime range, select indexes for the corresponding collection ID and datetime range, and then compile the final list of selected indexes:

`["sentinel-2-l2a"]:  {'gte': '2010-01-01T00:00:00.0Z', 'lte': '2025-12-31T23:59:59.0Z'}`
`["sentinel-2-l1c"]:  {'gte': '2012-01-01T00:00:00.0Z', 'lte': '2025-12-31T23:59:59.0Z'}`

jonhealy1 commented on CHANGELOG.md:
```diff
@@ -9,6 +9,10 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
+- Added CQL2 Abstract Syntax Tree (AST) structure for efficient query parsing and datetime-based indexes. [#560](https://github.com/stac-utils/stac-fastapi-elasticsearch-opensearch/pull/560)
+
+- Environment variable `VALIDATE_QUERYABLES` to enable/disable validation of queryables in search/filter requests. When set to `true`, search requests will be validated against the defined queryables, returning an error for any unsupported fields. Defaults to `false` for backward compatibility.[#532](https://github.com/stac-utils/stac-fastapi-elasticsearch-opensearch/pull/532)
```
Can you remove this entry for VALIDATE_QUERYABLES? This entry is already in the changelog under v6.8.0

jonhealy1 reviewed (approved):
Thanks @YuriZmytrakov
